contracts/SaveDAI.sol

```diff
@@ -153,55 +252,81 @@ contract SaveDAI is ISaveDAI, ERC20, Pausable, AccessControl {
     }
 
-    /**
-    * @notice This function will remove insurance and exchange your saveDAI for cDAI
-    * @param _amount The amount of saveDAI tokens to unbundle
-    */
+    /// @notice This function will remove insurance and exchange your saveDAI for cDAI
+    /// @param _amount The amount of saveDAI tokens to unbundle
     function withdrawForAsset(uint256 _amount)
         external
         override
     {
         require(!ocDai.hasExpired(), "ocDAI must not have expired");
+        require(farmerProxy[msg.sender] != address(0), 
+            "The user must have a farmer address");
+
+        // identify saveDAI contract's cDAI balance
+        uint256 initialcDaiBalance = cDai.balanceOf(address(this));
+
         // swap _amount of ocDAI on Uniswap for DAI
         uint256 daiTokens = _uniswapBuyDai(_amount);
 
-        // mint cDAI
-        uint256 cDAItokens = _mintCDai(daiTokens);
+        // mint cDAI - returns 0 when executed successfully
+        require(cDai.mint(daiTokens) == 0);
+
+        // identify saveDAI contract's updated cDAI balance
+        uint256 updatedcDaiBalance = cDai.balanceOf(address(this));
 
-        // transfer the sum of the newly minted cDAI with the original _amount
-        require(cDai.transfer(msg.sender, cDAItokens.add(_amount)));
+        // determine amount of cDAI minted
+        uint256 cDaiTokens = updatedcDaiBalance.sub(initialcDaiBalance);
+
+        // get the proxy address to transfer cDAI
+        address proxy = farmerProxy[msg.sender];
+
+        // transfer cDAI from SaveTokenFarmer
+        require(ISaveTokenFarmer(proxy).transfer(address(this), _amount));
+        
+        // transfer sum of newly minted cDAI with the original _amount in the SaveTokenFarmer
+        require(cDai.transfer(msg.sender, cDaiTokens.add(_amount)));
+        
         emit WithdrawForAsset(msg.sender, _amount);
         _burn(msg.sender, _amount);
     }
 
-    /**
-    * @notice This function will remove insurance and exchange your saveDAI for DAI
-    * @param _amount The amount of saveDAI tokens to unbundle
-    */
+    /// @notice This function will remove insurance and exchange your saveDAI for DAI
+    /// @param _amount The amount of saveDAI tokens to unbundle
     function withdrawForUnderlyingAsset(uint256 _amount)
         external
         override
     {
         require(!ocDai.hasExpired(), "ocDAI must not have expired");
+        require(farmerProxy[msg.sender] != address(0), 
+            "The user must have a farmer address");
 
         // identify saveDAI contract's DAI balance
-        uint256 initiaDaiBalance = dai.balanceOf(address(this));
+        uint256 initialDaiBalance = dai.balanceOf(address(this));
+
+        // get the proxy address to redeem cDAI from
+        address proxy = farmerProxy[msg.sender];
 
-        // Redeem returns 0 on success
-        require(cDai.redeem(_amount) == 0, "redeem function must execute successfully");
+        // transfer cDAI from SaveTokenFarmer
```

Comment:
@dougiebuckets rather then redeeming and transferring DAI to the saveDAI contract, could we pass in the msg.sender's address so the DAI is sent directly to their wallet?